PR: https://github.com/odoo/odoo/pull/222169

From: 08b62a4bbcc6f9a391b2cc00a621ef4c76100229
From: Arnold Moyaux
Date: 2025-08-09 04:39:25

Breaking data model changes scores: del:4 + add:0, change matches:
-    _inherit = "product.category"
-    property_account_creditor_price_difference_categ = fields.Many2one(
-    property_account_creditor_price_difference = fields.Many2one(
-    _inherit = 'stock.valuation.layer'

Total Changes: 769

[REF] stock_account: New inventory valuation

Functional:
* Automated handling of periodic valuation
* Closing report with managment of good delivered not invoiced and bill
  to receive
* Realtime valuation only at invoice/bill (no more at delivery/receipt)
* Removal of input/output accounts
* Automatic periodic/realtime closing
* Error management at closing (back dating, rounding issue,...)
* Back dating
* Inventory value is justify from last documents rather than the full
  history (fifo)

Technical:
* Removal of `stock.valuation.layer`
* value is stored on the `stock.move`
* value is compute base on accounting documents
* value priority order is: manual val > bill/inv > po/so > std price
* value could be use with a context at date to get the value at date
* `product.value` to log manual update from the user without real
  justification
* It doesn't try to replay and correct the past. Instead it focus on the
  current situation and match it

The new valuation aims to reduce the complexity of realtime accounting
and improve the simple periodic valuation.

The value for move in is always computed base on account.move/so/po or
standard price if none are available. In periodic, we compare the
balance of valuation account and the computed value of product. We
suggest the difference. In real time, the COGS are posted with the
know value of the goods at the time, they are post. The possible error
(back dating,...) are erased with the closing entries (same as periodic)

There is no more account for input and output. Instead in the closing
report, we show the po/so that are delivered but not invoiced. This way
the user can create the accrual for it. If you have good receipt not
billed. The inventory value will be greater than stock value. It result
by an entry that debit stock account and credit the variation account.
Then an accrual entry is posted to debit the variation account and
credit a bill to receive account.

There is a lot of skip test in this PR. We will try to remove them asap.
We decided to merge it in this state since it's already functional and
we can get feedbacks sooner.

Remaining work:
* Clean the test and remove all the skip (priority to module not in our
  scope)
* Analytic accounting
* Adapt report for cost of good sold.
* Accrual entries in report and closing account move

closes odoo/odoo#222169

Related: odoo/enterprise#91890
Related: odoo/upgrade#8189
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/purchase_stock/models/__init__.py
+++ b/addons/purchase_stock/models/__init__.py
@@ -13,4 +13,3 @@ from . import res_company
 from . import stock
 from . import stock_move
 from . import stock_rule
-from . import stock_valuation_layer

--- a/addons/purchase_stock/models/account_invoice.py
+++ b/addons/purchase_stock/models/account_invoice.py
@@ -11,19 +11,20 @@ class AccountMove(models.Model):
 
     def _stock_account_prepare_anglo_saxon_in_lines_vals(self):
         ''' Prepare values used to create the journal items (account.move.line) corresponding to the price difference
-         lines for vendor bills. It only concerns the quantities that have been delivered before the bill
+        lines for vendor bills. It only concerns the quantities that have been delivered before the bill
         Example:
         Buy a product having a cost of 9 and a supplier price of 10 and being a storable product and having a perpetual
         valuation in FIFO. Deliver the product and then post the bill. The vendor bill's journal entries looks like:
+
         Account                                     | Debit | Credit
         ---------------------------------------------------------------
-        101120 Stock Interim Account (Received)     | 10.0  |
+        101120 Stock Account                        | 10.0  |
         ---------------------------------------------------------------
         101100 Account Payable                      |       | 10.0
         ---------------------------------------------------------------
         This method computes values used to make two additional journal items:
         ---------------------------------------------------------------
-        101120 Stock Interim Account (Received)     |       | 1.0
+        101120 Stock Account                        |       | 1.0
         ---------------------------------------------------------------
         xxxxxx Expenses                             | 1.0   |
         ---------------------------------------------------------------
@@ -40,14 +41,14 @@ class AccountMove(models.Model):
             for line in move.invoice_line_ids:
                 # Filter out lines being not eligible for price difference.
                 # Moreover, this function is used for standard cost method only.
-                if not line._eligible_for_cogs() or line.product_id.cost_method != 'standard':
+                if not line._eligible_for_stock_account() or line.product_id.cost_method != 'standard':
                     continue
 
                 # Retrieve accounts needed to generate the price difference.
+
                 debit_pdiff_account = False
                 if line.product_id.cost_method == 'standard':
-                    debit_pdiff_account = line.product_id.property_account_creditor_price_difference \
-                        or line.product_id.categ_id.property_account_creditor_price_difference_categ
+                    debit_pdiff_account = line.product_id.categ_id.property_price_difference_account_id
                     debit_pdiff_account = move.fiscal_position_id.map_account(debit_pdiff_account)
                 else:
                     debit_pdiff_account = line.product_id.product_tmpl_id.get_product_accounts(fiscal_pos=move.fiscal_position_id)['expense']
@@ -112,53 +113,19 @@ class AccountMove(models.Model):
                     lines_vals_list.append(vals)
         return lines_vals_list
 
+    def button_draft(self):
+        return super().button_draft()
+
     def _post(self, soft=True):
         if not self.env.context.get('move_reverse_cancel'):
             self.env['account.move.line'].create(self._stock_account_prepare_anglo_saxon_in_lines_vals())
 
-        # Create correction layer and impact accounts if invoice price is different
-        stock_valuation_layers = self.env['stock.valuation.layer'].sudo()
-        valued_lines = self.env['account.move.line'].sudo()
-        for invoice in self:
-            if invoice.sudo().stock_valuation_layer_ids:
-                continue
-            if invoice.move_type in ('in_invoice', 'in_refund', 'in_receipt'):
-                valued_lines |= invoice.invoice_line_ids.filtered(
-                    lambda l: l.product_id and l.product_id.cost_method != 'standard')
-        if valued_lines:
-            svls, _amls = valued_lines._apply_price_difference()
-            stock_valuation_layers |= svls
-
-        for product, company in unique((svl.product_id, svl.company_id) for svl in stock_valuation_layers):
-            product = product.with_company(company.id)
-            if not product.uom_id.is_zero(product.quantity_svl):
-                product.sudo().with_context(disable_auto_svl=True).write({'standard_price': product.value_svl / product.quantity_svl})
-
-        for lot, company in unique((svl.lot_id, svl.company_id) for svl in stock_valuation_layers):
-            if not lot:
-                continue
-            lot = lot.with_company(company.id)
-            if not lot.product_id.uom_id.is_zero(lot.quantity_svl):
-                lot.sudo().with_context(disable_auto_svl=True).write({'standard_price': lot.value_svl / lot.quantity_svl})
-
-        posted = super(AccountMove, self.with_context(skip_cogs_reconciliation=True))._post(soft)
-
-        # The invoice reference is set during the super call
-        for layer in stock_valuation_layers:
-            description = f"{layer.account_move_line_id.move_id.display_name} - {layer.product_id.display_name}"
-            layer.description = description
-
-        if stock_valuation_layers:
-            stock_valuation_layers._validate_accounting_entries()
-
-        self._stock_account_anglo_saxon_reconcile_valuation()
-
-        return posted
+        return super()._post(soft)
 
     def _stock_account_get_last_step_stock_moves(self):
         """ Overridden from stock_account.
         Returns the stock moves associated to this invoice."""
-        rslt = super(AccountMove, self)._stock_account_get_last_step_stock_moves()
+        rslt = super()._stock_account_get_last_step_stock_moves()
         for invoice in self.filtered(lambda x: x.move_type == 'in_invoice'):
             rslt += invoice.mapped('invoice_line_ids.purchase_line_id.move_ids').filtered(lambda x: x.state == 'done' and x.location_id.usage == 'supplier')
         for invoice in self.filtered(lambda x: x.move_type == 'in_refund'):

--- a/addons/purchase_stock/models/account_move_line.py
+++ b/addons/purchase_stock/models/account_move_line.py
@@ -9,358 +9,20 @@ from collections import defaultdict
 class AccountMoveLine(models.Model):
     _inherit = 'account.move.line'
 
-    def _get_valued_in_moves(self):
-        self.ensure_one()
-        return self.purchase_line_id.move_ids.filtered(
-            lambda m: m.state == 'done' and m.product_qty != 0)
-
-    def _get_out_and_not_invoiced_qty(self, in_moves):
-        self.ensure_one()
-        if not in_moves:
-            return 0
-        aml_qty = self.product_uom_id._compute_quantity(self.quantity, self.product_id.uom_id)
-        invoiced_qty = sum(line.product_uom_id._compute_quantity(line.quantity, line.product_id.uom_id)
-                           for line in self.purchase_line_id.invoice_lines - self)
-        layers = in_moves.stock_valuation_layer_ids
-        layers_qty = sum(layers.mapped('quantity'))
-        out_qty = layers_qty - sum(layers.mapped('remaining_qty'))
-        total_out_and_not_invoiced_qty = max(0, out_qty - invoiced_qty)
-        out_and_not_invoiced_qty = min(aml_qty, total_out_and_not_invoiced_qty)
-        return self.product_id.uom_id._compute_quantity(out_and_not_invoiced_qty, self.product_uom_id)
-
-    def _apply_price_difference(self):
-        svl_vals_list = []
-        aml_vals_list = []
-        for line in self:
-            line = line.with_company(line.company_id)
-            po_line = line.purchase_line_id
-            uom = line.product_uom_id or line.product_id.uom_id
-
-            # Don't create value for more quantity than received
-            quantity = po_line.qty_received - (po_line.qty_invoiced - line.quantity)
-            quantity = max(min(line.quantity, quantity), 0)
-            if uom.is_zero(quantity):
-                continue
-
-            layers = line._get_valued_in_moves().stock_valuation_layer_ids.filtered(lambda svl: svl.product_id == line.product_id and not svl.stock_valuation_layer_id)
-            if not layers:
-                continue
-
-            new_svl_vals_list, new_aml_vals_list = line._generate_price_difference_vals(layers)
-            svl_vals_list += new_svl_vals_list
-            aml_vals_list += new_aml_vals_list
-        return self.env['stock.valuation.layer'].sudo().create(svl_vals_list), self.env['account.move.line'].sudo().create(aml_vals_list)
-
-    def _generate_price_difference_vals(self, layers):
-        """
-        The method will determine which layers are impacted by the AML (`self`) and, in case of a price difference, it
-        will then return the values of the new AMLs and SVLs
-        """
-        self.ensure_one()
-        po_line = self.purchase_line_id
-        product_uom = self.product_id.uom_id
-
-        # `history` is a list of tuples: (time, aml, layer)
-        # aml and layer will never be both defined
-        # we use this to get an order between posted AML and layers
-        history = [(layer.create_date, False, layer) for layer in layers]
-        am_state_field = self.env['ir.model.fields'].search([('model', '=', 'account.move'), ('name', '=', 'state')], limit=1)
-        for aml in po_line.invoice_lines:
-            move = aml.move_id
-            if move.state != 'posted':
-                continue
-            state_trackings = move.message_ids.tracking_value_ids.filtered(lambda t: t.field_id == am_state_field).sorted('id')
-            time = state_trackings[-1:].create_date or move.create_date  # `or` in case it has been created in posted state
-            history.append((time, aml, False))
-        # Sort history based on the datetime. In case of equality, the prority is given to SVLs, then to IDs.
-        # That way, we ensure a deterministic behaviour
-        history.sort(key=lambda item: (item[0], bool(item[1]), (item[1] or item[2]).id))
-
-        # Replay the whole history: we want to know what are the links between each layer and each invoice,
-        # and then the links between `self` and the layers
-        qty_to_invoice_per_layer, layers_and_invoices_qties = self._replay_history(layers, history)
-
-        # Now we know what layers does `self` use, let's check if we have to create a pdiff SVL
-        # (or cancel such an SVL in case of a refund)
-        invoice = self.move_id
-        svl_vals_list = []
-        aml_vals_list = []
-        for layer in layers:
-            # use the link between `self` and `layer` (i.e. the qty of `layer` billed by `self`)
-            invoicing_layer_qty = layers_and_invoices_qties[(layer, invoice)][1]
-            if product_uom.is_zero(invoicing_layer_qty):
-                continue
-
-            # We will only consider the total quantity to invoice of the layer because we don't
-            # want to invoice a part of the layer that has not been invoiced and that has been
-            # returned in the meantime
-            total_layer_qty_to_invoice = qty_to_invoice_per_layer[layer][0]
-            remaining_qty = layer.remaining_qty
-            out_layer_qty = total_layer_qty_to_invoice - remaining_qty
-            if self.is_refund:
-                sign = -1
-                reversed_invoice = invoice.reversed_entry_id
-                if not reversed_invoice:
-                    # this is a refund for a returned quantity, we don't have anything to do
-                    continue
-                initial_invoiced_qty = layers_and_invoices_qties[(layer, reversed_invoice)][0]
-                initial_pdiff_svl = layer.stock_valuation_layer_ids.filtered(lambda svl: svl.account_move_line_id.move_id == reversed_invoice)
-                if not initial_pdiff_svl or product_uom.is_zero(initial_invoiced_qty):
-                    continue
-                # We have an already-out quantity: we must skip the part already invoiced. So, first,
-                # let's compute the already invoiced quantity...
-                previously_invoiced_qty = 0
-                for item in history:
-                    previous_aml = item[1]
-                    if not previous_aml or previous_aml.is_refund:
-                        continue
-                    previous_invoice = previous_aml.move_id
-                    if previous_invoice == reversed_invoice:
-                        break
-                    previously_invoiced_qty += layers_and_invoices_qties[(layer, previous_invoice,)][1]
-                # ... Second, skip it:
-                out_qty_to_invoice = max(0, out_layer_qty - previously_invoiced_qty)
-                qty_to_correct = max(0, invoicing_layer_qty - out_qty_to_invoice)
-                if out_qty_to_invoice:
-                    # In case the out qty is different from the one posted by the initial bill, we should compensate
-                    # this quantity with debit/credit between stock_in and expense, but we are reversing an initial
-                    # invoice and don't want to do more than the original one
-                    out_qty_to_invoice = 0
-                aml = initial_pdiff_svl.account_move_line_id
-                parent_layer = initial_pdiff_svl.stock_valuation_layer_id
-                layer_price_unit = parent_layer._get_layer_price_unit()
-            else:
-                sign = 1
-                # get the invoiced qty of the layer without considering `self`
-                invoiced_layer_qty = total_layer_qty_to_invoice - qty_to_invoice_per_layer[layer][1] - invoicing_layer_qty
-                remaining_out_qty_to_invoice = max(0, out_layer_qty - invoiced_layer_qty)
-                out_qty_to_invoice = min(remaining_out_qty_to_invoice, invoicing_layer_qty)
-                qty_to_correct = invoicing_layer_qty - out_qty_to_invoice
-                layer_price_unit = layer._get_layer_price_unit()
-
-                returned_move = layer.stock_move_id.origin_returned_move_id
-                if returned_move and returned_move._is_out() and returned_move._is_returned(valued_type='out'):
-                    # Odd case! The user receives a product, then returns it. The returns are processed as classic
-                    # output, so the value of the returned product can be different from the initial one. The user
-                    # then receives again the returned product (that's where we are here) -> the SVL is based on
-                    # the returned one, the accounting entries are already compensated, and we don't want to impact
-                    # the stock valuation. So, let's fake the layer price unit with the POL one as everything is
-                    # already ok
-                    layer_price_unit = po_line.currency_id._convert(
-                        po_line._get_gross_price_unit(),
-                        layer.currency_id,
-                        layer.company_id,
-                        layer.create_date.date(),
-                        round=False
-                    )
-                aml = self
-
-            svl_vals, aml_vals = self._prepare_pdiff_vals(layer, aml, layer_price_unit, out_qty_to_invoice, sign * qty_to_correct)
-            svl_vals_list.extend(svl_vals)
-            aml_vals_list.extend(aml_vals)
-
-        return svl_vals_list, aml_vals_list
-
-    def _replay_history(self, layers, history):
-        history.append((False, self, False))  # time was only usefull for the sorting
-
-        # the next dict is a matrix [layer L, invoice I] where each cell gives two info:
-        # [initial qty of L invoiced by I, remaining invoiced qty]
-        # the second info is usefull in case of a refund
-        layers_and_invoices_qties = defaultdict(lambda: [0, 0])
-
-        # the next dict will also provide two info:
-        # [total qty to invoice, remaining qty to invoice]
-        # we need the total qty to invoice, so we will be able to deduce the invoiced qty before `self`
-        qty_to_invoice_per_layer = defaultdict(lambda: [0, 0])
-        for _time, aml, layer in history:
-            if layer:
-                total_layer_qty_to_invoice = abs(layer.quantity)
-                initial_layer = layer.stock_move_id.origin_returned_move_id.stock_valuation_layer_ids
-                # Filter out revaluation layers (Landed Cost)
-                initial_layer = initial_layer.filtered(lambda svl: not svl.stock_valuation_layer_id)
-                if initial_layer:
-                    # `layer` is a return. We will cancel the qty to invoice of the returned layer
-                    # /!\ we will cancel the qty not yet invoiced only
-                    initial_layer_remaining_qty = qty_to_invoice_per_layer[initial_layer][1]
-                    common_qty = min(initial_layer_remaining_qty, total_layer_qty_to_invoice)
-                    qty_to_invoice_per_layer[initial_layer][0] -= common_qty
-                    qty_to_invoice_per_layer[initial_layer][1] -= common_qty
-                    total_layer_qty_to_invoice = max(0, total_layer_qty_to_invoice - common_qty)
-                if self.product_id.uom_id.compare(total_layer_qty_to_invoice, 0) > 0:
-                    qty_to_invoice_per_layer[layer] = [total_layer_qty_to_invoice, total_layer_qty_to_invoice]
-            else:
-                invoice = aml.move_id
-                impacted_invoice = False
-                aml_qty = aml.product_uom_id._compute_quantity(aml.quantity, self.product_id.uom_id)
-                if aml.is_refund:
-                    reversed_invoice = aml.move_id.reversed_entry_id
-                    if reversed_invoice:
-                        sign = -1
-                        impacted_invoice = reversed_invoice
-                        # it's a refund, therefore we can only consume the quantities invoiced by
-                        # the initial invoice (`reversed_invoice`)
-                        layers_to_consume = []
-                        for layer in layers:
-                            remaining_invoiced_qty = layers_and_invoices_qties[layer, reversed_invoice][1]
-                            layers_to_consume.append((layer, remaining_invoiced_qty))
-                    else:
-                        # the refund has been generated because of a stock return, let's find and use it
-                        sign = 1
-                        layers_to_consume = []
-                        for layer in qty_to_invoice_per_layer:
-                            if layer.stock_move_id._is_out():
-                                layers_to_consume.append((layer, qty_to_invoice_per_layer[layer][1]))
-                else:
-                    # classic case, we are billing a received quantity so let's use the incoming SVLs
-                    sign = 1
-                    layers_to_consume = []
-                    for layer in qty_to_invoice_per_layer:
-                        if layer.stock_move_id._is_in():
-                            layers_to_consume.append((layer, qty_to_invoice_per_layer[layer][1]))
-                while self.product_id.uom_id.compare(aml_qty, 0) > 0 and layers_to_consume:
-                    layer, total_layer_qty_to_invoice = layers_to_consume[0]
-                    layers_to_consume = layers_to_consume[1:]
-                    if self.product_id.uom_id.is_zero(total_layer_qty_to_invoice):
-                        continue
-                    common_qty = min(aml_qty, total_layer_qty_to_invoice)
-                    aml_qty -= common_qty
-                    qty_to_invoice_per_layer[layer][1] -= sign * common_qty
-                    layers_and_invoices_qties[layer, invoice] = [common_qty, common_qty]
-                    layers_and_invoices_qties[layer, impacted_invoice][1] -= common_qty
-
-        return qty_to_invoice_per_layer, layers_and_invoices_qties
-
-    def _prepare_pdiff_vals(self, layer, aml, layer_price_unit, out_qty_to_invoice, qty_to_correct):
-        svl_vals_list = []
-        aml_vals_list = []
-
-        aml_gross_price_unit = aml._get_gross_unit_price()
-        # convert from aml currency to company currency
-        aml_price_unit = aml_gross_price_unit / aml.currency_rate
-        aml_price_unit = aml.product_uom_id._compute_price(aml_price_unit, self.product_id.uom_id)
-
-        unit_valuation_difference = aml_price_unit - layer_price_unit
-        precision_digits = max(
-            aml.currency_id.decimal_places,
-            layer.currency_id.decimal_places,
-            self.env['decimal.precision'].precision_get('Product Price'),
-        )
-        if float_is_zero(unit_valuation_difference, precision_digits=precision_digits):
-            unit_valuation_difference = 0
-
-        # Generate the AML values for the already out quantities
-        # convert from company currency to aml currency
-        unit_valuation_difference_curr = unit_valuation_difference * self.currency_rate
-        unit_valuation_difference_curr = self.product_id.uom_id._compute_price(unit_valuation_difference_curr, self.product_uom_id)
-        out_qty_to_invoice = self.product_id.uom_id._compute_quantity(out_qty_to_invoice, self.product_uom_id)
-        if (
-            not self.currency_id.is_zero(unit_valuation_difference_curr * out_qty_to_invoice) and
-            self.product_id.valuation == 'real_time'
-        ):
-            aml_vals_list += self._prepare_pdiff_aml_vals(out_qty_to_invoice, unit_valuation_difference_curr)
-
-        # Generate the SVL values for the on hand quantities (and impact the parent layer)
-        po_pu_curr = self.purchase_line_id.currency_id._convert(
-            self.purchase_line_id.price_unit,
-            self.currency_id,
-            self.company_id,
-            self.move_id.invoice_date or self.date or fields.Date.context_today(self),
-            round=False
-        )
-        price_difference_curr = po_pu_curr - aml_gross_price_unit
-        if not float_is_zero(unit_valuation_difference * qty_to_correct, precision_rounding=self.company_id.currency_id.rounding):
-            svl_vals = self._prepare_pdiff_svl_vals(layer, qty_to_correct, unit_valuation_difference, price_difference_curr)
-            layer.remaining_value += svl_vals['value']
-            svl_vals_list.append(svl_vals)
-
-        return svl_vals_list, aml_vals_list
-
-    def _prepare_pdiff_aml_vals(self, qty, unit_valuation_difference):
-        self.ensure_one()
-        vals_list = []
-
-        sign = self.move_id.direction_sign
-        expense_account = self.product_id.product_tmpl_id.get_product_accounts(fiscal_pos=self.move_id.fiscal_position_id)['expense']
-        if not expense_account:
-            return vals_list
-
-        for price, account in [
-            (unit_valuation_difference, expense_account),
-            (-unit_valuation_difference, self.account_id),
-        ]:
-            vals_list.append({
-                'name': self.name[:64],
-                'move_id': self.move_id.id,
-                'partner_id': self.partner_id.id or self.move_id.commercial_partner_id.id,
-                'currency_id': self.currency_id.id,
-                'product_id': self.product_id.id,
-                'product_uom_id': self.product_uom_id.id,
-                'balance': self.company_id.currency_id.round((qty * price * sign) / self.currency_rate),
-                'account_id': account.id,
-                'analytic_distribution': self.analytic_distribution,
-                'display_type': 'cogs',
-            })
-        return vals_list
-
-    def _prepare_pdiff_svl_vals(self, corrected_layer, quantity, unit_cost, pdiff):
-        self.ensure_one()
-        common_svl_vals = {
-            'account_move_id': self.move_id.id,
-            'account_move_line_id': self.id,
-            'company_id': self.company_id.id,
-            'product_id': self.product_id.id,
-            'quantity': 0,
-            'unit_cost': 0,
-            'remaining_qty': 0,
-            'remaining_value': 0,
-            'description': self.move_id.name and '%s - %s' % (self.move_id.name, self.product_id.name) or self.product_id.name,
-        }
-        return {
-            **self.product_id._prepare_in_svl_vals(quantity, unit_cost, corrected_layer.lot_id),
-            **common_svl_vals,
-            'stock_valuation_layer_id': corrected_layer.id,
-            'price_diff_value': self.currency_id.round(pdiff * quantity),
-        }
-
     def _get_price_unit_val_dif_and_relevant_qty(self):
         self.ensure_one()
-        # Retrieve stock valuation moves.
-        valuation_stock_moves = self.env['stock.move'].search([
-            ('purchase_line_id', '=', self.purchase_line_id.id),
-            ('state', '=', 'done'),
-            ('product_qty', '!=', 0.0),
-        ]) if self.purchase_line_id else self.env['stock.move']
-
-        if self.product_id.cost_method != 'standard' and self.purchase_line_id:
-            if self.move_type == 'in_refund':
-                valuation_stock_moves = valuation_stock_moves.filtered(lambda stock_move: stock_move._is_out())
-            else:
-                valuation_stock_moves = valuation_stock_moves.filtered(lambda stock_move: stock_move._is_in())
 
-            if not valuation_stock_moves:
-                return 0, 0
-
-            valuation_price_unit_total, valuation_total_qty = valuation_stock_moves._get_valuation_price_and_qty(self, self.move_id.currency_id)
-            valuation_price_unit = valuation_price_unit_total / valuation_total_qty
-            valuation_price_unit = self.product_id.uom_id._compute_price(valuation_price_unit, self.product_uom_id)
-        else:
-            # Valuation_price unit is always expressed in invoice currency, so that it can always be computed with the good rate
-            price_unit = self.product_id.uom_id._compute_price(self.product_id.standard_price, self.product_uom_id)
-            price_unit = -price_unit if self.move_id.move_type == 'in_refund' else price_unit
-            valuation_date = valuation_stock_moves and max(valuation_stock_moves.mapped('date')) or self.date
-            valuation_price_unit = self.company_currency_id._convert(
-                price_unit, self.currency_id,
-                self.company_id, valuation_date, round=False
-            )
+        # Valuation_price unit is always expressed in invoice currency, so that it can always be computed with the good rate
+        valuation_price_unit = self.product_id.uom_id._compute_price(self.product_id.standard_price, self.product_uom_id)
+        valuation_price_unit = -valuation_price_unit if self.move_id.move_type == 'in_refund' else valuation_price_unit
 
         price_unit = self._get_gross_unit_price()
 
         price_unit_val_dif = price_unit - valuation_price_unit
         # If there are some valued moves, we only consider their quantity already used
-        if self.product_id.cost_method == 'standard':
-            relevant_qty = self.quantity
-        else:
-            relevant_qty = self._get_out_and_not_invoiced_qty(valuation_stock_moves)
+        relevant_qty = self.quantity
 
         return price_unit_val_dif, relevant_qty
+
+    def _get_stock_moves(self):
+        return super()._get_stock_moves() | self.purchase_line_id.move_ids

--- a/addons/purchase_stock/models/product.py
+++ b/addons/purchase_stock/models/product.py
@@ -7,23 +7,9 @@ from odoo.fields import Domain
 from odoo.tools import formatLang
 
 
-class ProductCategory(models.Model):
-    _inherit = "product.category"
-
-    property_account_creditor_price_difference_categ = fields.Many2one(
-        'account.account', string="Price Difference Account",
-        company_dependent=True, ondelete='restrict',
-        help="This account will be used to value price difference between purchase price and accounting cost.")
-
-
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    property_account_creditor_price_difference = fields.Many2one(
-        'account.account', string="Price Difference Account", company_dependent=True, ondelete='restrict',
-        help="This account is used in automated inventory valuation to "\
-             "record the price difference between a purchase order and its related vendor bill when validating this vendor bill.")
-
     @api.model
     def _get_buy_route(self):
         buy_route = self.env.ref('purchase_stock.route_warehouse0_buy', raise_if_not_found=False)

--- a/addons/purchase_stock/models/purchase_order_line.py
+++ b/addons/purchase_stock/models/purchase_order_line.py
@@ -87,7 +87,7 @@ class PurchaseOrderLine(models.Model):
 
     @api.model_create_multi
     def create(self, vals_list):
-        lines = super(PurchaseOrderLine, self).create(vals_list)
+        lines = super().create(vals_list)
         lines.filtered(lambda l: l.order_id.state == 'purchase')._create_or_update_picking()
         return lines
 
@@ -101,7 +101,7 @@ class PurchaseOrderLine(models.Model):
 
         previous_product_uom_qty = {line.id: line.product_uom_qty for line in lines}
         previous_product_qty = {line.id: line.product_qty for line in lines}
-        result = super(PurchaseOrderLine, self).write(values)
+        result = super().write(values)
         if 'price_unit' in values:
             for line in lines:
                 # Avoid updating kit components' stock.move
@@ -110,6 +110,9 @@ class PurchaseOrderLine(models.Model):
         if 'product_qty' in values:
             lines = lines.filtered(lambda l: l.product_uom_id.compare(previous_product_qty[l.id], l.product_qty) != 0)
             lines.with_context(previous_product_qty=previous_product_uom_qty)._create_or_update_picking()
+        valuation_trigger = ['price_unit', 'product_qty', 'product_uom']
+        if any(field in valuation_trigger for field in values):
+            self.move_ids.filtered(lambda m: m.is_valued)._set_value()
         return result
 
     def action_product_forecast_report(self):
@@ -193,30 +196,6 @@ class PurchaseOrderLine(models.Model):
                 moves = line._create_stock_moves(picking)
                 moves._action_confirm()._action_assign()
 
-    def _get_stock_move_price_unit(self):
-        self.ensure_one()
-        order = self.order_id
-        price_unit = self.price_unit
-        price_unit_prec = self.env['decimal.precision'].precision_get('Product Price')
-        if self.tax_ids:
-            qty = self.product_qty or 1
-            price_unit = self.tax_ids.compute_all(
-                price_unit,
-                currency=self.order_id.currency_id,
-                quantity=qty,
-                product=self.product_id,
-                partner=self.order_id.partner_id,
-                rounding_method="round_globally",
-            )['total_void']
-            price_unit = price_unit / qty
-        if self.product_uom_id.id != self.product_id.uom_id.id:
-            price_unit /= self.product_uom_id.factor
-            price_unit *= self.product_id.uom_id.factor
-        if order.currency_id != order.company_id.currency_id:
-            price_unit = order.currency_id._convert(
-                price_unit, order.company_id.currency_id, self.company_id, self.date_order or fields.Date.today(), round=False)
-        return float_round(price_unit, precision_digits=price_unit_prec)
-
     def _get_move_dests_initial_demand(self, move_dests):
         return self.product_id.uom_id._compute_quantity(
             sum(move_dests.filtered(lambda m: m.state != 'cancel' and m.location_dest_id.usage != 'supplier').mapped('product_qty')),
@@ -255,6 +234,30 @@ class PurchaseOrderLine(models.Model):
             res.append(extra_move_vals)
         return res
 
+    def _get_stock_move_price_unit(self):
+        self.ensure_one()
+        order = self.order_id
+        price_unit = self.price_unit
+        price_unit_prec = self.env['decimal.precision'].precision_get('Product Price')
+        if self.tax_ids:
+            qty = self.product_qty or 1
+            price_unit = self.tax_ids.compute_all(
+                price_unit,
+                currency=self.order_id.currency_id,
+                quantity=qty,
+                product=self.product_id,
+                partner=self.order_id.partner_id,
+                rounding_method="round_globally",
+            )['total_void']
+            price_unit = price_unit / qty
+        if self.product_uom_id.id != self.product_id.uom_id.id:
+            price_unit /= self.product_uom_id.factor
+            price_unit *= self.product_id.uom_id.factor
+        if order.currency_id != order.company_id.currency_id:
+            price_unit = order.currency_id._convert(
+                price_unit, order.company_id.currency_id, self.company_id, self.date_order or fields.Date.today(), round=False)
+        return float_round(price_unit, precision_digits=price_unit_prec)
+
     def _get_qty_procurement(self):
         self.ensure_one()
         qty = 0.0

--- a/addons/purchase_stock/models/stock_move.py
+++ b/addons/purchase_stock/models/stock_move.py
@@ -56,208 +56,6 @@ class StockMove(models.Model):
     def _get_description(self):
         return self.purchase_line_id.name if self.purchase_line_id else super()._get_description()
 
-    def _should_ignore_pol_price(self):
-        self.ensure_one()
-        return self.origin_returned_move_id or not self.purchase_line_id or not self.product_id.id
-
-    def _get_price_unit(self):
-        """ Returns the unit price for the move"""
-        self.ensure_one()
-        if self._should_ignore_pol_price():
-            return super(StockMove, self)._get_price_unit()
-        price_unit_prec = self.env['decimal.precision'].precision_get('Product Price')
-        line = self.purchase_line_id
-        order = line.order_id
-        received_qty = self._get_qty_received_without_self()
-        if line.product_id.purchase_method == 'purchase' and line.product_uom_id.compare(line.qty_invoiced, received_qty) > 0:
-            move_layer = line.move_ids.sudo().stock_valuation_layer_ids
-            invoiced_layer = line.sudo().invoice_lines.stock_valuation_layer_ids
-            # value on valuation layer is in company's currency, while value on invoice line is in order's currency
-            receipt_value = 0
-            for layer in move_layer:
-                if not layer._should_impact_price_unit_receipt_value():
-                    continue
-                receipt_value += layer.currency_id._convert(
-                    layer.value, order.currency_id, order.company_id, layer.create_date, round=False)
-            if invoiced_layer:
-                receipt_value += sum(invoiced_layer.mapped(lambda l: l.currency_id._convert(
-                    l.value, order.currency_id, order.company_id, l.create_date, round=False)))
-            total_invoiced_value = 0
-            invoiced_qty = 0
-            for invoice_line in line.sudo().invoice_lines:
-                if invoice_line.move_id.state != 'posted':
-                    continue
-                # Adjust unit price to account for discounts before adding taxes.
-                adjusted_unit_price = invoice_line.price_unit * (1 - (invoice_line.discount / 100)) if invoice_line.discount else invoice_line.price_unit
-                if invoice_line.tax_ids:
-                    invoice_line_value = invoice_line.tax_ids.compute_all(
-                        adjusted_unit_price,
-                        currency=invoice_line.currency_id,
-                        quantity=invoice_line.quantity,
-                        rounding_method="round_globally",
-                    )['total_void']
-                else:
-                    invoice_line_value = adjusted_unit_price * invoice_line.quantity
-                total_invoiced_value += invoice_line.currency_id._convert(
-                        invoice_line_value, order.currency_id, order.company_id, invoice_line.move_id.invoice_date, round=False)
-                invoiced_qty += invoice_line.product_uom_id._compute_quantity(invoice_line.quantity, line.product_id.uom_id, rounding_method="HALF-UP")
-            # TODO currency check
-            remaining_value = total_invoiced_value - receipt_value
-            # TODO qty_received in product uom
-            remaining_qty = invoiced_qty - line.product_uom_id._compute_quantity(received_qty, line.product_id.uom_id, rounding_method="HALF-UP")
-            has_remaining = (
-                not order.currency_id.is_zero(remaining_value)
-                and not float_is_zero(remaining_qty, precision_rounding=line.product_id.uom_id.rounding)
-            )
-            if order.currency_id != order.company_id.currency_id and has_remaining:
-                # will be rounded during currency conversion
-                price_unit = remaining_value / remaining_qty
-            elif has_remaining:
-                price_unit = float_round(remaining_value / remaining_qty, precision_digits=price_unit_prec)
-            else:
-                price_unit = line._get_gross_price_unit()
-        else:
-            price_unit = line._get_gross_price_unit()
-        if order.currency_id != order.company_id.currency_id:
-            convert_date = self._get_currency_convert_date()
-            price_unit = order.currency_id._convert(
-                price_unit, order.company_id.currency_id, order.company_id, convert_date, round=False)
-        if self.product_id.lot_valuated:
-            return dict.fromkeys(self.lot_ids, price_unit)
-        return {self.env['stock.lot']: price_unit}
-
-    def _get_qty_received_without_self(self):
-        qty_received = self.purchase_line_id.qty_received
-        if self.state == 'done':
-            qty_received -= self.product_uom._compute_quantity(
-                self.quantity, self.purchase_line_id.product_uom_id, rounding_method='HALF-UP'
-            )
-        return qty_received
-
-    def _get_currency_convert_date(self):
-        self.ensure_one()
-        # The date must be today, and not the date of the move since the move move is still
-        # in assigned state. However, the move date is the scheduled date until move is
-        # done, then date of actual move processing. See:
-        # https://github.com/odoo/odoo/blob/2f789b6863407e63f90b3a2d4cc3be09815f7002/addons/stock/models/stock_move.py#L36
-        convert_date = fields.Date.context_today(self) if self.state != 'done' else self.date
-        line = self.purchase_line_id
-        if not line:
-            return convert_date
-
-        # Use currency rate at bill date when invoice before receipt
-        qty_received = self._get_qty_received_without_self()
-        if line.product_uom_id.compare(line.qty_invoiced, qty_received) > 0:
-            posted_bills = line.sudo().invoice_lines.move_id.filtered(lambda m: m.state == 'posted')
-            convert_date = max(posted_bills.mapped('invoice_date'), default=convert_date)
-        return convert_date
-
-    def _generate_valuation_lines_data(self, partner_id, qty, debit_value, credit_value, debit_account_id, credit_account_id, svl_id, description):
-        """ Overridden from stock_account to support amount_currency on valuation lines generated from po
-        """
-        self.ensure_one()
-
-        rslt = super(StockMove, self)._generate_valuation_lines_data(partner_id, qty, debit_value, credit_value, debit_account_id, credit_account_id, svl_id, description)
-        purchase_currency = self.purchase_line_id.currency_id
-        company_currency = self.company_id.currency_id
-        if not self.purchase_line_id or purchase_currency == company_currency:
-            return rslt
-        svl = self.env['stock.valuation.layer'].browse(svl_id)
-        if not svl.account_move_line_id:
-            convert_date = self._get_currency_convert_date()
-            rslt['credit_line_vals']['amount_currency'] = company_currency._convert(
-                rslt['credit_line_vals']['balance'],
-                purchase_currency,
-                self.company_id,
-                convert_date
-            )
-            rslt['debit_line_vals']['amount_currency'] = company_currency._convert(
-                rslt['debit_line_vals']['balance'],
-                purchase_currency,
-                self.company_id,
-                convert_date
-            )
-            rslt['debit_line_vals']['currency_id'] = purchase_currency.id
-            rslt['credit_line_vals']['currency_id'] = purchase_currency.id
-        else:
-            rslt['credit_line_vals']['amount_currency'] = 0
-            rslt['debit_line_vals']['amount_currency'] = 0
-            rslt['debit_line_vals']['currency_id'] = purchase_currency.id
-            rslt['credit_line_vals']['currency_id'] = purchase_currency.id
-            if not svl.price_diff_value:
-                return rslt
-            # The idea is to force using the company currency during the reconciliation process
-            rslt['debit_line_vals_curr'] = {
-                'name': _("Currency exchange rate difference"),
-                'product_id': self.product_id.id,
-                'quantity': 0,
-                'product_uom_id': self.product_id.uom_id.id,
-                'partner_id': partner_id,
-                'balance': 0,
-                'account_id': debit_account_id,
-                'currency_id': purchase_currency.id,
-                'amount_currency': -svl.price_diff_value,
-            }
-            rslt['credit_line_vals_curr'] = {
-                'name': _("Currency exchange rate difference"),
-                'product_id': self.product_id.id,
-                'quantity': 0,
-                'product_uom_id': self.product_id.uom_id.id,
-                'partner_id': partner_id,
-                'balance': 0,
-                'account_id': credit_account_id,
-                'currency_id': purchase_currency.id,
-                'amount_currency': svl.price_diff_value,
-            }
-        return rslt
-
-    def _account_entry_move(self, qty, description, svl_id, cost):
-        """
-        In case of a PO return, if the value of the returned product is
-        different from the purchased one, we need to empty the stock_in account
-        with the difference
-        """
-        move_directions = self.env.context.get('move_directions') or False
-        am_vals_list = super(StockMove, self.with_context(move_directions=move_directions))._account_entry_move(qty, description, svl_id, cost)
-        returned_move = self.origin_returned_move_id
-        move = (self | returned_move).with_prefetch(self._prefetch_ids)
-        pdiff_exists = bool(move.stock_valuation_layer_ids.stock_valuation_layer_ids.account_move_line_id)
-
-        if not am_vals_list or not self.purchase_line_id or pdiff_exists or self.product_id.uom_id.is_zero(qty):
-            return am_vals_list
-
-        layer = self.env['stock.valuation.layer'].browse(svl_id)
-
-        self_is_out_move = False
-        if move_directions:
-            self_is_out_move = move_directions.get(self.id) and 'out' in move_directions.get(self.id)
-        else:
-            self_is_out_move = self._is_out()
-
-        if returned_move and self_is_out_move and self._is_returned(valued_type='out'):
-            returned_layer = returned_move.stock_valuation_layer_ids.filtered(lambda svl: not svl.stock_valuation_layer_id)[:1]
-            unit_diff = layer._get_layer_price_unit() - returned_layer._get_layer_price_unit() if returned_layer else 0
-        elif returned_move and returned_move._is_out() and returned_move._is_returned(valued_type='out'):
-            returned_layer = returned_move.stock_valuation_layer_ids.filtered(lambda svl: not svl.stock_valuation_layer_id)[:1]
-            unit_diff = returned_layer._get_layer_price_unit() - self.purchase_line_id._get_gross_price_unit()
-        else:
-            return am_vals_list
-
-        diff = unit_diff * qty
-        company = self.purchase_line_id.company_id
-        if company.currency_id.is_zero(diff):
-            return am_vals_list
-
-        sm = self.with_company(company).with_context(is_returned=True)
-        accounts = sm.product_id.product_tmpl_id.get_product_accounts()
-        acc_exp_id = accounts['expense'].id
-        acc_stock_in_id = accounts['stock_input'].id
-        journal_id = accounts['stock_journal'].id
-        vals = sm._prepare_account_move_vals(acc_exp_id, acc_stock_in_id, journal_id, qty, description, False, diff)
-        am_vals_list.append(vals)
-
-        return am_vals_list
-
     def _prepare_move_split_vals(self, uom_qty):
         vals = super(StockMove, self)._prepare_move_split_vals(uom_qty)
         vals['purchase_line_id'] = self.purchase_line_id.id
@@ -279,49 +77,14 @@ class StockMove(models.Model):
         else:
             return super(StockMove, self)._get_upstream_documents_and_responsibles(visited)
 
-    def _get_related_invoices(self):
-        """ Overridden to return the vendor bills related to this stock move.
-        """
-        rslt = super(StockMove, self)._get_related_invoices()
-        purchase_ids = self.env['purchase.order'].search([('picking_ids', 'in', self.picking_id.ids)])
-        rslt += purchase_ids.invoice_ids.filtered(lambda x: x.state == 'posted')
-        return rslt
-
     def _get_source_document(self):
         res = super()._get_source_document()
         return self.purchase_line_id.order_id or res
 
-    def _get_valuation_price_and_qty(self, related_aml, to_curr):
-        valuation_price_unit_total = 0
-        valuation_total_qty = 0
-        for val_stock_move in self:
-            # In case val_stock_move is a return move, its valuation entries have been made with the
-            # currency rate corresponding to the original stock move
-            valuation_date = val_stock_move.origin_returned_move_id.date or val_stock_move.date
-            svl = val_stock_move.with_context(active_test=False).mapped('stock_valuation_layer_ids').filtered(
-                lambda l: l.quantity)
-            layers_qty = sum(svl.mapped('quantity'))
-            layers_values = sum(svl.mapped('value'))
-            valuation_price_unit_total += related_aml.company_currency_id._convert(
-                layers_values, to_curr, related_aml.company_id, valuation_date, round=False,
-            )
-            valuation_total_qty += layers_qty
-        if related_aml.product_uom_id.rounding or related_aml.product_id.uom_id.is_zero(valuation_total_qty):
-            raise UserError(
-                _('Odoo is not able to generate the anglo saxon entries. The total valuation of %s is zero.',
-                  related_aml.product_id.display_name))
-        return valuation_price_unit_total, valuation_total_qty
-
     def _is_purchase_return(self):
         self.ensure_one()
         return self.location_dest_id.usage == "supplier" or (self.origin_returned_move_id and self.location_dest_id == self.env.ref('stock.stock_location_inter_company', raise_if_not_found=False))
 
-    def _get_all_related_aml(self):
-        # The back and for between account_move and account_move_line is necessary to catch the
-        # additional lines from a cogs correction
-        return super()._get_all_related_aml() | self.purchase_line_id.invoice_lines.move_id.line_ids.filtered(
-            lambda aml: aml.product_id == self.purchase_line_id.product_id)
-
     def _get_all_related_sm(self, product):
         return super()._get_all_related_sm(product) | self.filtered(lambda m: m.purchase_line_id.product_id == product)
 
@@ -337,3 +100,45 @@ class StockMove(models.Model):
                 [move for move in current_move.move_orig_ids if move not in moves_to_check and move not in seen_moves]
             )
         return None, None
+
+    # --------------------------------------------------------
+    # Valuation
+    # --------------------------------------------------------
+
+    def _get_value_from_account_move(self, quantity, at_date=None):
+        if not (self.purchase_line_id and self.purchase_line_id):
+            return 0, 0
+
+        quantity = 0
+        value = 0
+        for aml in self.purchase_line_id.invoice_lines:
+            if at_date and aml.date > at_date:
+                continue
+            if aml.move_id.state != 'posted':
+                continue
+            if aml.move_type == 'in_invoice':
+                quantity += aml.quantity
+                value += aml.price_subtotal
+            elif aml.move_type == 'in_refund':
+                quantity -= aml.quantity
+                value -= aml.price_subtotal
+
+        return value, quantity
+
+    def _get_value_from_quotation(self, quantity, at_date=None):
+        # TODO: Start from global value
+        if not self.purchase_line_id:
+            return super()._get_value_from_quotation(quantity)
+        if at_date and self.purchase_line_id.order_id.date_order > at_date:
+            return super()._get_value_from_quotation(quantity)
+        price_unit = self.purchase_line_id._get_stock_move_price_unit()
+        quantity = min(quantity, self.quantity)
+        return price_unit * quantity, quantity
+
+    def _get_related_invoices(self):
+        """ Overridden to return the vendor bills related to this stock move.
+        """
+        rslt = super()._get_related_invoices()
+        purchase_ids = self.env['purchase.order'].search([('picking_ids', 'in', self.picking_id.ids)])
+        rslt += purchase_ids.invoice_ids.filtered(lambda x: x.state == 'posted')
+        return rslt

--- a/addons/purchase_stock/models/stock_valuation_layer.py
+++ b/None
@@ -1,11 +0,0 @@
-from odoo import models
-
-
-class StockValuationLayer(models.Model):
-    _inherit = 'stock.valuation.layer'
-
-    def _get_layer_price_unit(self):
-        """ This function returns the value of product in a layer per unit, relative to the aml
-            the function is designed to be overriden to add logic to price unit calculation
-        """
-        return self.value / self.quantity
